File: packages/raster-plugin-haptics/ios/Sources/HapticsPlugin/HapticsPlugin.swift
import RasterPluginSDK
import UIKit

public enum HapticsPlugin {
    public static func register() {
        RasterPlugin.register(plugin: "Haptics", method: "impact") { call in
            DispatchQueue.main.async {
                let style = (call.args?["style"] as? String) ?? "medium"
                let generator: UIImpactFeedbackGenerator
                switch style {
                case "light":
                    generator = UIImpactFeedbackGenerator(style: .light)
                case "heavy":
                    generator = UIImpactFeedbackGenerator(style: .heavy)
                default:
                    generator = UIImpactFeedbackGenerator(style: .medium)
                }
                generator.prepare()
                generator.impactOccurred()
                call.replyOk(["ok": true])
            }
        }

        RasterPlugin.register(plugin: "Haptics", method: "vibrate") { call in
            DispatchQueue.main.async {
                let generator = UINotificationFeedbackGenerator()
                generator.prepare()
                generator.notificationOccurred(.success)
                call.replyOk(["ok": true])
            }
        }
    }
}

File: packages/raster-plugin-haptics/android/src/main/java/dev/raster/plugin/haptics/HapticsPlugin.kt
package dev.raster.plugin.haptics

import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import dev.raster.plugin.RasterPlugin
import org.json.JSONObject

object HapticsPlugin {
    @JvmStatic
    fun register() {
        RasterPlugin.register("Haptics", "impact") { call ->
            val activity = RasterPlugin.currentActivity()
            if (activity == null) {
                call.replyErr("NO_ACTIVITY", "No activity")
                return@register
            }
            val style = call.args?.optString("style", "medium") ?: "medium"
            val duration = when (style) {
                "light" -> 20L
                "heavy" -> 60L
                else -> 40L
            }
            vibrate(activity, duration)
            call.replyOk(JSONObject().put("ok", true))
        }

        RasterPlugin.register("Haptics", "vibrate") { call ->
            val activity = RasterPlugin.currentActivity()
            if (activity == null) {
                call.replyErr("NO_ACTIVITY", "No activity")
                return@register
            }
            val duration = call.args?.optLong("duration", 50L) ?: 50L
            vibrate(activity, duration.coerceIn(1L, 500L))
            call.replyOk(JSONObject().put("ok", true))
        }
    }

    private fun vibrate(activity: android.app.Activity, durationMs: Long) {
        val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val manager = activity.getSystemService(VibratorManager::class.java)
            manager?.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            activity.getSystemService(Vibrator::class.java)
        } ?: return

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(durationMs, VibrationEffect.DEFAULT_AMPLITUDE))
        } else {
            @Suppress("DEPRECATION")
            vibrator.vibrate(durationMs)
        }
    }
}